### PR TITLE
Include py.typed file in distribution (PEP 561)

### PR DESCRIPTION
**Pull request by tlambert03**
## Description

In order for the `.pyi` files to be useful,  scikit-image needs to include a `py.typed` file in top level module of the distribution.

See [PEP 561](https://peps.python.org/pep-0561/) and the [mypy documentation](https://mypy.readthedocs.io/en/stable/installed_packages.html#creating-pep-561-compatible-packages) for more information.

currently, when using scikit-image in an IDE, you get:
Screen Shot 2023-07-21 at 7 48 08 PM

if you drop `py.typed` file in the skimage folder of site-packages you get:

Screen Shot 2023-07-21 at 7 48 44 PM

### CODE REVIEW

**Review by lagru — APPROVED** [review]
Thanks for catching this Talley! Seems obvious that we want to include this change. 

Though, I can't seem to reproduce this with PyCharm on my side; behavior is identical with and without `py.typed`. And PyCharm finds the docstring without the file.  Which IDE are you using, a vscode-based one?

**Comment by tlambert03** [comment]
Yep, vscode

**Comment by tlambert03** [comment]
If I recall correctly, pycharm actually does dynamic imports for a number of things?

**Review by jarrodmillman — APPROVED** [review]

**Comment by lagru** [comment]
Thanks @tlambert03. :)

**Comment by darosio** [comment]
I was used to a single:
import skimage  # type: ignore
or a few:
from skimage.morphology import disk  # type: ignore

Now I need a "# type: ignore" for every still untyped function I use to make the mypy check happy.
Which is fine if you are planning to add types to all/most functions in the near future. Is that the case? I could not find in the docs related planning.

**Comment by tlambert03** [comment]
You can use the mypy configuration to batch ignore, or selectively ignore specific errors (or all errors):
https://mypy.readthedocs.io/en/stable/config_file.html#suppressing-errors

**Comment by tlambert03** [comment]
for a little added context: 
the motivation here goes back to https://github.com/scikit-image/scikit-image/pull/5101 which introduced lazy submodule imports for many modules in sci-kit image.  Because modules were no longer listed as literal imports, type checkers and IDEs were no longer able to infer imports and namespaces for intellisense/autocomplete.

So, in order to tell IDEs what scikit-image has available, two options were presented (#6435 and #6429), and a variant of #6435 was merged in #6577, using pyi stubs to indicate to IDEs what modules & functions scikit-image had to offer.  However, in order for static type checkers to use these static hints, you must include a py.typed file (as per PEP 561).

My understanding of the implications of py.typed from PEP 561 is _not_ to indicate some critical mass or coverage of typed functions in a package, but simply to indicate to type checkers that they should search _this_ package for stubs, as opposed to a third party stubs package (or typeshed) which was the historical/legacy way that many packages incrementally offered types.   Users who wish to use mypy with strict setting (I'm definitely a fan) often do need to add a bit of configuration for partially typed dependencies.

I don't particularly have a preference here, i leave it to the maintainers of scikit-image to decide... but if you would vote for the removal of py.typed, then I think you also need to suggest how you would deal with the inability of IDEs to detect _anything_ in scikit-image given the use of lazy imports.

**Comment by johnthagen** [comment]
The problem is that the way the suppressions work, you can't add a blanket ignore for problems in a third party library (at least I could not get this to work) because the _calls_ to `skimage` happen in user code. If you have these calls scattered all over your code base, you end up having to suppress `disallow_untyped_calls` everywhere, which opens you up for the type checker missing real errors in other usages of your code.

For example, I tried:

```toml
[[tool.mypy.overrides]]
module = "skimage"
disallow_untyped_calls = false
```

But this does not suppress `skimage` issues:

```
my/package/module.py:1: error: Call to untyped function "img_as_ubyte" in typed context  [no-untyped-call]
```

I would be happy to be proven wrong if there is a Mypy configuration that can suppress _only_ `skimage`-caused `no-untyped-call` errors globally.

**Comment by soupault** [comment]
>if scikit-image ever wants to offer type hints for a broader portion of the library

I personally find type-hinting very useful, both in prototyping and proper development scenarios. So, in my mind, we should definitely go that path.

**Comment by johnthagen** [comment]
> so, it will be available soon

Thanks for the reference! That certainly will limit the impact of this change and allow users to suppress this particular issue. I agree that after the next Mypy release, the negative impacts of this PR will not be nearly as severe.

**Comment by stefanv** [comment]
It sounds like the problem will be resolved by (a) a new feature now available in mypy and (b) our stated attempt to type annotate and implement protocols as part of skimage v2.

Perhaps @lagru and @jni have more to say about that plan.

**Comment by lagru** [comment]
@johnthagen, sorry for the py.typed related annoyance from our side. 

We currently have a grant that also entails typing our API. It's a big effort and we are also trying to revise our API in the process but you can follow along in 
- https://github.com/scikit-image/scikit-image/pull/7185
- https://github.com/scikit-image/scikit-image/pull/6905
- https://github.com/scikit-image/scikit-image/wiki/API-changes-for-skimage2

From skimming the discussion above, our best bet seems to be to continue work on that and leave the py.typed be.

**Comment by johnthagen** [comment]
As of Mypy 1.6.1, you can now suppress _only_ the untyped call errors in `skimage` with:

```toml
[tool.mypy]
strict = true
untyped_calls_exclude = [
    "skimage"
]
```

**Pull request merged**
